File: src/trl/rng.rs
use super::opt::Opt;
use ::core::marker::Sized;
use ::core::ops::Fn;

pub trait Rng {
    type Key;
    type Item;

    fn left(&self) -> Opt<Self::Item>;
    fn right(&self) -> Opt<Self::Item>;
    fn popLeft(&mut self) -> (Self, Opt<Self::Item>)
    where
        Self: Sized;
    fn popRight(&mut self) -> (Self, Opt<Self::Item>)
    where
        Self: Sized;
    fn pop(&mut self) -> (Self, Opt<Self::Item>)
    where
        Self: Sized;
    fn at(&self, key: Self::Key) -> Opt<Self::Item>;

    fn len(&self) -> usize;

    fn every<F>(&self, f: F) -> bool
    where
        Self: Sized,
        F: Fn(Self::Item, Self::Key) -> bool;

    fn some<F>(&self, f: F) -> bool
    where
        Self: Sized,
        F: Fn(Self::Item, Self::Key) -> bool;

    fn fold<F, R>(&self, initial_value: R, f: F) -> R
    where
        Self: Sized,
        F: Fn(R, Self::Item, usize) -> R;

    fn reduce<F>(&self, f: F) -> Self::Item
    where
        Self: Sized,
        F: Fn(Self::Item, Self::Item, usize) -> Self::Item;

    fn map<F, R>(&self, f: F) -> dyn Rng<Key = Self::Key, Item = R>
    where
        Self: Sized,
        F: Fn(Self::Item, Self::Key) -> R;

    fn filter<F>(&self, f: F) -> Self
    where
        Self: Sized,
        F: Fn(Self::Item, Self::Key) -> bool;

    fn find<F>(&self, f: F) -> Opt<(Self::Item, Self::Key)>
    where
        Self: Sized,
        F: Fn(Self::Item, Self::Key) -> bool;

    fn includes(&self, item: Self::Item) -> bool;

    fn each<F>(&self, f: F)
    where
        Self: Sized,
        F: Fn(Self::Item, Self::Key);
}

File: src/trl/rng.ts
import { Opt } from "./opt"

export interface Rng<K, T> {
    left(): Opt<T>
    right(): Opt<T>
    popLeft(): [Rng<K, T>, Opt<T>]
    popRight(): [Rng<K, T>, Opt<T>]
    pop(): [Rng<K, T>, Opt<T>]
    at(key: K): Opt<T>
    len(): usize

    every<F>(f: (T, K) => bool): bool
    some<F>(f: (T, K) => bool): bool
    fold<F, R>(initial_value: R, f: (R, T, usize) => R): R
    reduce<F>(f: (T, T, usize) => T): T
    map<F, R>(f: (T, K) => R): dyn Rng<Key = K, Item = R>
    filter<F>(f: (T, K) => bool): Self
    find<F>(f: (T, K) => bool): Opt<(T, K)>
    includes(item: T): bool
    each<F>(f: (T, K) => void)
}
